src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -137,6 +137,37 @@ struct document_context {
 // incompatibilities don't matter, only whether they are compatible or not
 static const std::filesystem::path ignored_path = "";
 
+// helper struct to hold a json::Document or a json::Value::ConstObject, used in
+// ref_resolution to either hold an existing json::Value from the root object or
+// a synthesized one, when a $ref has siblings
+struct json_const_object {
+    std::variant<json::Value::ConstObject, std::unique_ptr<json::Document>>
+      storage;
+    json_const_object(const json::Value::ConstObject& co) noexcept
+      : storage(co) {}
+    json_const_object(std::unique_ptr<json::Document>&& d) noexcept
+      : storage(std::move(d)) {}
+
+    operator const json::Value&() const {
+        return std::visit(
+          ss::make_visitor(
```

Comment:
note: `ss::visit` should return `decltype(auto)`

Reply:
yes, I always forget that but that's the reason why I default to std::visit(ss::make_visitor (that and the support for multiple variants)